// wargamer/lib/modules/WorldOfWarships/Encyclopedia.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _fuse = require('fuse.js');

var _fuse2 = _interopRequireDefault(_fuse);

var _ClientModule2 = require('../ClientModule');

var _ClientModule3 = _interopRequireDefault(_ClientModule2);

var _Encyclopedia = require('../mixins/Encyclopedia');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * @classdesc Module for the World of Warships Encyclopedia endpoint.
 * @extends ClientModule
 */
var Encyclopedia = function (_ClientModule) {
  _inherits(Encyclopedia, _ClientModule);

  /**
   * Constructor.
   * @param {BaseClient} client - The API client this module belongs to.
   */
  function Encyclopedia(client) {
    _classCallCheck(this, Encyclopedia);

    /**
     * The module's Fuse object.
     * @type {Fuse}
     * @private
     */
    var _this = _possibleConstructorReturn(this, (Encyclopedia.__proto__ || Object.getPrototypeOf(Encyclopedia)).call(this, client, 'encyclopedia'));

    _this.fuse = new _fuse2.default([], {
      keys: ['name']
    });
    return _this;
  }

  /**
   * Searches for a ship by name or ID and returns its entry from the
   *   `encyclopedia/ships` endpoint.
   * @param {(number|string)} identifier - The ship identifier to use for lookup.
   * If a number is supplied, it is treated as the ship's ID.
   * If a string is supplied, the identifier is matched against ship names with
   *   the closest match being selected.
   * @returns {Promise.<?Object, Error>} A promise resolving to the data for the
   *   matched ship, or `null` if no ships were matched.
   */


  _createClass(Encyclopedia, [{
    key: 'findShip',
    value: function findShip(identifier) {
      return _Encyclopedia.resolveEntry.call(this, {
        identifier,
        indexEndpoint: 'encyclopedia/ships',
        dataEndpoint: 'encyclopedia/ships',
        identifierKey: 'ship_id',
        fuse: this.fuse,
        searchFields: ['name']
      });
    }

    /**
     * Localizes a ship type slug.
     * @param {string} slug - The slug.
     * @returns {Promise.<(string|undefined), Error>} Promise resolving to the
     *   translated slug, or `undefined` if it couldn't be translated.
     */

  }, {
    key: 'localizeShipType',
    value: function localizeShipType(slug) {
      return _Encyclopedia.localize.call(this, {
        method: 'encyclopedia/info',
        type: 'ship_types',
        slug
      });
    }

    /**
     * Localizes a language slug.
     * @param {string} slug - The slug.
     * @returns {Promise.<(string|undefined), Error>} Promise resolving to the
     *   translated slug, or `undefined` if it couldn't be translated.
     */

  }, {
    key: 'localizeLanguage',
    value: function localizeLanguage(slug) {
      return _Encyclopedia.localize.call(this, {
        method: 'encyclopedia/info',
        type: 'languages',
        slug
      });
    }

    /**
     * Localizes a ship modification slug.
     * @param {string} slug - The slug.
     * @returns {Promise.<(string|undefined), Error>} Promise resolving to the
     *   translated slug, or `undefined` if it couldn't be translated.
     */

  }, {
    key: 'localizeShipModification',
    value: function localizeShipModification(slug) {
      return _Encyclopedia.localize.call(this, {
        method: 'encyclopedia/info',
        type: 'ship_modifications',
        slug
      });
    }

    /**
     * Localizes a ship module slug.
     * @param {string} slug - The slug.
     * @returns {Promise.<(string|undefined), Error>} Promise resolving to the
     *   translated slug, or `undefined` if it couldn't be translated.
     */

  }, {
    key: 'localizeShipModule',
    value: function localizeShipModule(slug) {
      return _Encyclopedia.localize.call(this, {
        method: 'encyclopedia/info',
        type: 'ship_modules',
        slug
      });
    }

    /**
     * Localizes a ship nation slug.
     * @param {string} slug - The slug.
     * @returns {Promise.<(string|undefined), Error>} Promise resolving to the
     *   translated slug, or `undefined` if it couldn't be translated.
     */

  }, {
    key: 'localizeShipNation',
    value: function localizeShipNation(slug) {
      return _Encyclopedia.localize.call(this, {
        method: 'encyclopedia/info',
        type: 'ship_nations',
        slug
      });
    }
  }]);

  return Encyclopedia;
}(_ClientModule3.default);

exports.default = Encyclopedia;
