// wargamer/lib/modules/WorldOfTanks/Tankopedia.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _fuse = require('fuse.js');

var _fuse2 = _interopRequireDefault(_fuse);

var _ClientModule2 = require('../ClientModule');

var _ClientModule3 = _interopRequireDefault(_ClientModule2);

var _Encyclopedia = require('../mixins/Encyclopedia');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * @classdesc Module for the World of Tanks Tankopedia endpoint.
 * @extends ClientModule
 */
var Tankopedia = function (_ClientModule) {
  _inherits(Tankopedia, _ClientModule);

  /**
   * Constructor.
   * @param {BaseClient} client - The API client this module belongs to.
   */
  function Tankopedia(client) {
    _classCallCheck(this, Tankopedia);

    /**
     * Mapping between module types and their ID field names.
     * @type {Object}
     * @static
     * @const
     * @private
     */
    var _this = _possibleConstructorReturn(this, (Tankopedia.__proto__ || Object.getPrototypeOf(Tankopedia)).call(this, client, 'tankopedia'));

    _this.constructor.MODULE_ID_FIELDS = {
      vehicleChassis: 'suspension_id',
      vehicleEngine: 'engine_id',
      vehicleGun: 'gun_id',
      vehicleRadio: 'radio_id',
      vehicleTurret: 'turret_id'
    };

    /**
     * The module's Fuse object.
     * @type {Fuse}
     * @private
     */
    _this.fuse = new _fuse2.default([], {
      keys: ['name', 'short_name']
    });
    return _this;
  }

  /**
   * Searches for a vehicle by name or ID and returns its entry from the
   *   `encyclopedia/vehicles` endpoint.
   * @param {(number|string)} identifier - The vehicle identifier to use for
   *   lookup.
   * If a number is supplied, it is treated as the vehicle's ID.
   * If a string is supplied, the identifier is matched against vehicle names
   *   with the closest match being selected.
   * @returns {Promise.<?Object, Error>} A promise resolving to the data for the
   *   matched vehicle, or `null` if no vehicles were matched.
   */


  _createClass(Tankopedia, [{
    key: 'findVehicle',
    value: function findVehicle(identifier) {
      return _Encyclopedia.resolveEntry.call(this, {
        identifier,
        indexEndpoint: 'encyclopedia/vehicles',
        dataEndpoint: 'encyclopedia/vehicles',
        identifierKey: 'tank_id',
        fuse: this.fuse,
        searchFields: ['name', 'short_name']
      });
    }

    /**
     * Returns the vehicle profile for a given vehicle.
     * @param {number} vehicleId - The vehicle ID to use for lookup.
     * @param {string} [profile='stock'] - The vehicle profile to lookup. Can be
     *   a profile ID, or one of `'stock'` or `'top'`. The top configuration is
     *   determined by picking the most expensive modules to research in the
     *   vehicle's research tree.
     * @returns {Promise.<?Object, Error>} A promise resolving to the data for the
     *   matched vehicle profile, or `null` if no vehicle was matched.
     */

  }, {
    key: 'findVehicleProfile',
    value: function findVehicleProfile(vehicleId) {
      var _this2 = this;

      var profile = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'stock';

      if (profile === 'stock') {
        return this.client.get('encyclopedia/vehicleprofile', { tank_id: vehicleId }).then(function (response) {
          return response.data[vehicleId];
        });
      } else if (profile === 'top') {
        return this.findVehicle(vehicleId).then(function (vehicleData) {
          if (!vehicleData) {
            return null;
          }

          var topModules = (0, _Encyclopedia.extractTopModules)(vehicleData.modules_tree);
          var queryFields = Object.keys(topModules).reduce(function (accumulated, next) {
            var fieldName = _this2.constructor.MODULE_ID_FIELDS[next];

            return _extends({}, accumulated, {
              [fieldName]: topModules[next].module_id
            });
          }, {});

          return _this2.client.get('encyclopedia/vehicleprofile', _extends({}, queryFields, { tank_id: vehicleId }));
        }).then(function (result) {
          return result && result.data[vehicleId];
        });
      }

      return this.client.get('encyclopedia/vehicleprofile', { tank_id: vehicleId, profile_id: profile }).then(function (response) {
        return response.data[vehicleId];
      });
    }

    /**
     * Localizes a crew role slug.
     * @param {string} slug - The slug.
     * @returns {Promise.<(string|undefined), Error>} Promise resolving to the
     *   translated slug, or `undefined` if it couldn't be translated.
     */

  }, {
    key: 'localizeCrewRole',
    value: function localizeCrewRole(slug) {
      return _Encyclopedia.localize.call(this, {
        method: 'encyclopedia/info',
        type: 'vehicle_crew_roles',
        slug
      });
    }

    /**
     * Localizes a language slug.
     * @param {string} slug - The slug.
     * @returns {Promise.<(string|undefined), Error>} Promise resolving to the
     *   translated slug, or `undefined` if it couldn't be translated.
     */

  }, {
    key: 'localizeLanguage',
    value: function localizeLanguage(slug) {
      return _Encyclopedia.localize.call(this, {
        method: 'encyclopedia/info',
        type: 'languages',
        slug
      });
    }

    /**
     * Localizes an achievement section slug. The returned value is the section's
     *   name.
     * @param {string} slug - The slug.
     * @returns {Promise.<(string|undefined), Error>} Promise resolving to the
     *   translated slug, or `undefined` if it couldn't be translated.
     */

  }, {
    key: 'localizeAchievementSection',
    value: function localizeAchievementSection(slug) {
      return _Encyclopedia.localize.call(this, {
        method: 'encyclopedia/info',
        type: 'achievement_sections',
        slug
      }).then(function (section) {
        return section && section.name;
      });
    }

    /**
     * Localizes a vehicle type slug.
     * @param {string} slug - The slug.
     * @returns {Promise.<(string|undefined), Error>} Promise resolving to the
     *   translated slug, or `undefined` if it couldn't be translated.
     */

  }, {
    key: 'localizeVehicleType',
    value: function localizeVehicleType(slug) {
      return _Encyclopedia.localize.call(this, {
        method: 'encyclopedia/info',
        type: 'vehicle_types',
        slug
      });
    }

    /**
     * Localizes a vehicle nation slug.
     * @param {string} slug - The slug.
     * @returns {Promise.<(string|undefined), Error>} Promise resolving to the
     *   translated slug, or `undefined` if it couldn't be translated.
     */

  }, {
    key: 'localizeVehicleNation',
    value: function localizeVehicleNation(slug) {
      return _Encyclopedia.localize.call(this, {
        method: 'encyclopedia/info',
        type: 'vehicle_nations',
        slug
      });
    }
  }]);

  return Tankopedia;
}(_ClientModule3.default);

exports.default = Tankopedia;
