// wargamer/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Wargaming = exports.WorldOfWarplanes = exports.WorldOfWarships = exports.WorldOfTanksConsole = exports.WorldOfTanksBlitz = exports.WorldOfTanks = undefined;

var _Wargamer = require('./Wargamer');

var _Wargamer2 = _interopRequireDefault(_Wargamer);

var _WorldOfTanks = require('./clients/WorldOfTanks');

var _WorldOfTanks2 = _interopRequireDefault(_WorldOfTanks);

var _WorldOfTanksBlitz = require('./clients/WorldOfTanksBlitz');

var _WorldOfTanksBlitz2 = _interopRequireDefault(_WorldOfTanksBlitz);

var _WorldOfTanksConsole = require('./clients/WorldOfTanksConsole');

var _WorldOfTanksConsole2 = _interopRequireDefault(_WorldOfTanksConsole);

var _WorldOfWarships = require('./clients/WorldOfWarships');

var _WorldOfWarships2 = _interopRequireDefault(_WorldOfWarships);

var _WorldOfWarplanes = require('./clients/WorldOfWarplanes');

var _WorldOfWarplanes2 = _interopRequireDefault(_WorldOfWarplanes);

var _Wargaming = require('./clients/Wargaming');

var _Wargaming2 = _interopRequireDefault(_Wargaming);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.WorldOfTanks = _WorldOfTanks2.default;
exports.WorldOfTanksBlitz = _WorldOfTanksBlitz2.default;
exports.WorldOfTanksConsole = _WorldOfTanksConsole2.default;
exports.WorldOfWarships = _WorldOfWarships2.default;
exports.WorldOfWarplanes = _WorldOfWarplanes2.default;
exports.Wargaming = _Wargaming2.default;
exports.default = _Wargamer2.default;
