// wargamer/lib/Wargamer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _WorldOfTanks = require('./clients/WorldOfTanks');

var _WorldOfTanks2 = _interopRequireDefault(_WorldOfTanks);

var _WorldOfTanksBlitz = require('./clients/WorldOfTanksBlitz');

var _WorldOfTanksBlitz2 = _interopRequireDefault(_WorldOfTanksBlitz);

var _WorldOfTanksConsole = require('./clients/WorldOfTanksConsole');

var _WorldOfTanksConsole2 = _interopRequireDefault(_WorldOfTanksConsole);

var _WorldOfWarships = require('./clients/WorldOfWarships');

var _WorldOfWarships2 = _interopRequireDefault(_WorldOfWarships);

var _WorldOfWarplanes = require('./clients/WorldOfWarplanes');

var _WorldOfWarplanes2 = _interopRequireDefault(_WorldOfWarplanes);

var _Wargaming = require('./clients/Wargaming');

var _Wargaming2 = _interopRequireDefault(_Wargaming);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * @classdesc The Wargamer client.
 */
var Wargamer = function () {
  function Wargamer() {
    _classCallCheck(this, Wargamer);
  }

  _createClass(Wargamer, null, [{
    key: 'WoT',

    /**
     * Constructs a new World of Tanks API client.
     * @param {ClientOptions} options - The client options.
     * @returns {WorldOfTanks} The API client.
     * @static
     */
    value: function WoT(options) {
      return new _WorldOfTanks2.default(options);
    }

    /**
     * Constructs a new World of Tanks Blitz API client.
     * @param {ClientOptions} options - The client options.
     * @returns {WorldOfTanksBlitz} The API client.
     * @static
     */

  }, {
    key: 'WoTB',
    value: function WoTB(options) {
      return new _WorldOfTanksBlitz2.default(options);
    }

    /**
     * Constructs a new World of Tanks Console API client.
     * @param {ClientOptions} options - The client options.
     * @returns {WorldOfTanksConsole} The API client.
     * @static
     */

  }, {
    key: 'WoTX',
    value: function WoTX(options) {
      return new _WorldOfTanksConsole2.default(options);
    }

    /**
     * Constructs a new World of Warships API client.
     * @param {ClientOptions} options - The client options.
     * @returns {WorldOfWarships} The API client.
     * @static
     */

  }, {
    key: 'WoWS',
    value: function WoWS(options) {
      return new _WorldOfWarships2.default(options);
    }

    /**
     * Constructs a new World of Warplanes API client.
     * @param {ClientOptions} options - The client options.
     * @returns {WorldOfWarplanes} The API client.
     * @static
     */

  }, {
    key: 'WoWP',
    value: function WoWP(options) {
      return new _WorldOfWarplanes2.default(options);
    }

    /**
     * Constructs a new Wargaming.net API client.
     * @param {ClientOptions} options - The client options.
     * @returns {Wargaming} The API client.
     * @static
     */

  }, {
    key: 'WGN',
    value: function WGN(options) {
      return new _Wargaming2.default(options);
    }
  }]);

  return Wargamer;
}();

exports.default = Wargamer;
